packages/medusa/src/services/totals-new.ts

```diff
@@ -0,0 +1,692 @@
+   * @param calculationContext
+   */
+  async getLineItemTotals(
+    item: LineItem,
+    {
+      includeTax,
+      lineItemAllocation,
+      taxLines,
+      calculationContext,
+    }: {
+      includeTax?: boolean
+      lineItemAllocation: LineAllocationsMap[number]
+      taxLines?: (LineItemTaxLine | ShippingMethodTaxLine)[]
+      calculationContext: TaxCalculationContext
+    }
+  ): Promise<LineItemTotals> {
+    let subtotal = item.unit_price * item.quantity
+    if (
+      this.featureFlagRouter_.isFeatureEnabled(
+        TaxInclusivePricingFeatureFlag.key
+      ) &&
+      item.includes_tax
+    ) {
+      subtotal = 0 // in that case we need to know the tax rate to compute it later
+    }
+
+    const discount_total =
+      (lineItemAllocation.discount?.unit_amount || 0) * item.quantity
+
+    const totals: LineItemTotals = {
+      unit_price: item.unit_price,
+      quantity: item.quantity,
+      subtotal,
+      discount_total,
+      total: subtotal - discount_total,
+      original_total: subtotal,
+      original_tax_total: 0,
+      tax_total: 0,
+      tax_lines: (taxLines ?? item.tax_lines ?? []) as LineItemTaxLine[],
+    }
+
+    if (item.is_return) {
+      if (typeof item.tax_lines === "undefined") {
+        throw new MedusaError(
+          MedusaError.Types.UNEXPECTED_STATE,
+          "Return Line Items must join tax lines"
+        )
+      }
+    }
+
+    // Force the tax lines to exist anyway
+    if (includeTax && !totals.tax_lines.length) {
+      throw new MedusaError(
+        MedusaError.Types.UNEXPECTED_STATE,
+        "Tax Lines must be joined to calculate taxes"
+      )
+    }
+
+    if (totals.tax_lines.length > 0) {
+      totals.tax_total = await this.taxCalculationStrategy_.calculate(
+        [item],
+        totals.tax_lines,
+        calculationContext
+      )
+      const noDiscountContext = {
+        ...calculationContext,
+        allocation_map: {}, // Don't account for discounts
+      }
+
+      totals.original_tax_total = await this.taxCalculationStrategy_.calculate(
+        [item],
+        totals.tax_lines,
+        noDiscountContext
+      )
+
+      if (
+        this.featureFlagRouter_.isFeatureEnabled(
+          TaxInclusivePricingFeatureFlag.key
+        ) &&
+        item.includes_tax
+      ) {
+        totals.subtotal +=
+          totals.unit_price * totals.quantity - totals.original_tax_total
+        totals.total += totals.subtotal
+        totals.original_total += totals.subtotal
+      }
+
+      totals.total += totals.tax_total
+      totals.original_total += totals.original_tax_total
+    }
+
+    return totals
+  }
+
+  /**
+   * Calcul and return the legacy calculated totals using the tax rate
+   * @param item
+   * @param taxRate
+   * @param lineItemAllocation
+   * @param calculationContext
+   */
+  async getLineItemTotalsLegacy(
```

Comment:
thought: can't we delete this altogether as this is a "new" service?

Reply:
I kept it thinking that we might still want to benefit from the improvement even if you use a legacy tax system. Unless we say that we don't support those type of object that use the tax rate anymore. The idea is that the actual totals service will be replaced by the new one once we have refactored everything

Reply:
Ah alright - sorry thought this was the old method from the existing TotalsService

Reply:
ahah no problem, none of the method from this service are the same of the old one, they have been re worked 👍